src/unit_tests.zig

```diff
@@ -46,43 +29,160 @@ comptime {
     _ = @import("testing/marks.zig");
-    _ = @import("testing/storage.zig");
     _ = @import("testing/table.zig");
-    _ = @import("testing/tmp_tigerbeetle.zig");
-
+    _ = @import("tidy.zig");
+    _ = @import("time.zig");
+    _ = @import("trace.zig");
+    _ = @import("trace/event.zig");
     _ = @import("vsr.zig");
-    _ = @import("vsr/client.zig");
-    _ = @import("vsr/clock.zig");
     _ = @import("vsr/checksum.zig");
-    _ = @import("vsr/grid_blocks_missing.zig");
+    _ = @import("vsr/clock.zig");
+    _ = @import("vsr/free_set.zig");
     _ = @import("vsr/grid_scrubber.zig");
     _ = @import("vsr/journal.zig");
     _ = @import("vsr/marzullo.zig");
+    _ = @import("vsr/multi_batch.zig");
     _ = @import("vsr/replica_format.zig");
-    _ = @import("vsr/replica_reformat.zig");
     _ = @import("vsr/replica_test.zig");
     _ = @import("vsr/routing.zig");
     _ = @import("vsr/superblock.zig");
-    _ = @import("vsr/free_set.zig");
     _ = @import("vsr/superblock_quorums.zig");
-    _ = @import("vsr/sync.zig");
-    _ = @import("vsr/multi_batch.zig");
+}
 
-    _ = @import("scripts/release.zig");
-    _ = @import("scripts/changelog.zig");
-    _ = @import("scripts/cfo.zig");
+const std = @import("std");
+const builtin = @import("builtin");
+const assert = std.debug.assert;
+const build_root = @import("build_root").build_root;
+const max_source_size = 1024 * 1024;
+
+test {
+    var arena_instance = std.heap.ArenaAllocator.init(std.testing.allocator);
+    defer arena_instance.deinit();
+    const arena = arena_instance.allocator();
+
+    const src_path = try std.fmt.allocPrint(
+        arena,
+        "{s}/{s}",
+        .{ build_root, "src" },
+    );
+    var src_dir = try std.fs.openDirAbsolute(src_path, .{
+        .access_sub_paths = true,
+        .iterate = true,
+    });
+
+    var unit_tests_contents = std.ArrayList(u8).init(arena);
+    try unit_tests_contents.writer().writeAll("comptime {\n");
+
+    for (try unit_test_files(arena, src_dir)) |unit_test_file| {
+        try unit_tests_contents.writer().print("    _ = @import(\"{s}\");\n", .{unit_test_file});
+    }
+
+    try unit_tests_contents.writer().writeAll("}\n");
+
+    assert(std.mem.eql(u8, @src().file, "unit_tests.zig"));
+    const unit_tests_contents_disk = try src_dir.readFileAlloc(
+        arena,
+        @src().file,
+        max_source_size,
+    );
+    assert(std.mem.startsWith(u8, unit_tests_contents_disk, "comptime {"));
+    assert(std.mem.endsWith(u8, unit_tests_contents.items, "}\n"));
+
+    const unit_tests_needs_update = !std.mem.startsWith(
+        u8,
+        unit_tests_contents_disk,
+        unit_tests_contents.items,
+    );
+
+    if (unit_tests_needs_update) {
+        if (std.process.hasEnvVarConstant("SNAP_UPDATE")) {
+            const includes_end = std.mem.indexOf(u8, unit_tests_contents_disk, "}\n").? + 2;
+
+            // Add the rest of the real file on disk to the generated in-memory file.
```

Comment:
I am wondering if we should turn this into a true quine which doesn't read its own source code, just for cool points 🙈